File: chapters/1-memory-access/c/memory-access.c
#include <emscripten.h>
#include <stdlib.h>

int arr[10] = {1, 2, 3, 4, 5};

EMSCRIPTEN_KEEPALIVE
int sum(int a, int b) {
	return a + b;
}

EMSCRIPTEN_KEEPALIVE
int* getOffset() {
	return &arr[0];
}

typedef struct {
	int x;
	int y;
	char *name;
} Point;


EMSCRIPTEN_KEEPALIVE
Point* getPoint() {
	Point *p = malloc(sizeof(Point));
	p->x = 88323;
	p->y = 20;
	p->name = "hello";

	return p;
}

EMSCRIPTEN_KEEPALIVE
char* inverseImg(char* img, int size) {
	int i;
	for (i = 0; i < size; i++) {
		img[i] = 255 - img[i];
	}

	return img;
}

EMSCRIPTEN_KEEPALIVE
Point* createPoint(int x, int y, char* name) {
	Point *p = malloc(sizeof(Point));
	p->x = x;
	p->y = y;
	p->name = name;

	return p;
}


int main() {
   // intentionally left empty
}

File: chapters/1-memory-access/rust/memory-access.rs
#[no_mangle]
pub fn simple_arr() -> *mut i32 {
	let vec = Box::new([100, 200, 300, 400]);

	Box::into_raw(vec) as *mut i32
}

#[no_mangle]
pub extern fn get_offset() -> *mut i32 {
	let vec = &mut [1, 2, 3, 4];

	return vec.as_mut_ptr();
}

#[no_mangle]
pub extern fn sum(a: i32, b: i32) -> i32 {
	a + b
}

fn main() {
	// Intentionally left empty
}